/*
 * @begin:license
 *
 * Copyright (c) 2015-2019, Benjamin Niemann <pink@odahoda.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * @end:license
 */

syntax = "proto2";

package noisicaa.pb;

import "noisicaa/music/model_base.proto";
import "noisicaa/builtin_nodes/score_track/model.proto";
import "noisicaa/builtin_nodes/beat_track/model.proto";
import "noisicaa/builtin_nodes/control_track/model.proto";
import "noisicaa/builtin_nodes/sample_track/model.proto";
import "noisicaa/builtin_nodes/instrument/model.proto";
import "noisicaa/builtin_nodes/custom_csound/model.proto";
import "noisicaa/builtin_nodes/midi_source/model.proto";
import "noisicaa/builtin_nodes/step_sequencer/model.proto";
import "noisicaa/builtin_nodes/midi_cc_to_cv/model.proto";
import "noisicaa/builtin_nodes/midi_looper/model.proto";
import "noisicaa/builtin_nodes/midi_monitor/model.proto";
import "noisicaa/builtin_nodes/metronome/model.proto";
import "noisicaa/builtin_nodes/midi_velocity_mapper/model.proto";

extend ObjectBase {
  // Score track (401xxx)
  optional ScoreTrack score_track = 401000;
  optional ScoreMeasure score_measure = 401001;
  optional Note note = 401002;

  // Beat track (402xxx)
  optional BeatTrack beat_track = 402000;
  optional BeatMeasure beat_measure = 402001;
  optional Beat beat = 402002;

  // Control track (403xxx)
  optional ControlTrack control_track = 403000;
  optional ControlPoint control_point = 403001;

  // Sample track (404xxx)
  optional SampleTrack sample_track = 404000;
  optional SampleRef sample_ref = 404001;

  // Instrument (405xxx)
  optional Instrument instrument = 405000;

  // Custom CSound (407xxx)
  optional CustomCSound custom_csound = 407000;
  optional CustomCSoundPort custom_csound_port = 407001;

  // MIDI source (408xxx)
  optional MidiSource midi_source = 408000;

  // Step Sequencer (409xxx)
  optional StepSequencer step_sequencer = 409000;
  optional StepSequencerChannel step_sequencer_channel = 409001;
  optional StepSequencerStep step_sequencer_step = 409002;

  // MIDI CC to CV (410xxx)
  optional MidiCCtoCV midi_cc_to_cv = 410000;
  optional MidiCCtoCVChannel midi_cc_to_cv_channel = 410001;

  // MIDI Looper (411xxx)
  optional MidiLooper midi_looper = 411000;
  optional MidiLooperPatch midi_looper_patch = 411001;

  // MIDI Monitor (412xxx)
  optional MidiMonitor midi_monitor = 412000;

  // Metronome (413xxx)
  optional Metronome metronome = 413000;

  // MIDI Velocity Mapper (414xxx)
  optional MidiVelocityMapper midi_velocity_mapper = 414000;
}
